contracts/exchangeIssuance/ExchangeIssuanceV2.sol

```diff
@@ -0,0 +1,808 @@
+     * @param _tokenA       The address of tokenA
+     * @param _tokenB       The address of tokenB
+     *
+     * @return              The min amount of tokenA required across both exchanges
+     * @return              The Exchange on which minimum amount of tokenA is required
+     * @return              The pair address of the uniswap/sushiswap pool containing _tokenA and _tokenB
+     */
+    function _getMinTokenForExactToken(uint256 _amountOut, address _tokenA, address _tokenB) internal view returns (uint256, Exchange, address) {
+        if (_tokenA == _tokenB) {
+            return (_amountOut, Exchange.None, ETH_ADDRESS);
+        }
+
+        uint256 maxIn = PreciseUnitMath.maxUint256() ;
+        uint256 uniTokenIn = maxIn;
+        uint256 sushiTokenIn = maxIn;
+
+        address uniswapPair = _getPair(uniFactory, _tokenA, _tokenB);
+        if (uniswapPair != address(0)) {
+            (uint256 reserveIn, uint256 reserveOut) = UniSushiV2Library.getReserves(uniswapPair, _tokenA, _tokenB);
+            // Prevent subtraction overflow by making sure pool reserves are greater than swap amount
+            if (reserveOut > _amountOut) {
+                uniTokenIn = UniSushiV2Library.getAmountIn(_amountOut, reserveIn, reserveOut);
+            }
+        }
+
+        address sushiswapPair = _getPair(sushiFactory, _tokenA, _tokenB);
+        if (sushiswapPair != address(0)) {
+            (uint256 reserveIn, uint256 reserveOut) = UniSushiV2Library.getReserves(sushiswapPair, _tokenA, _tokenB);
+            // Prevent subtraction overflow by making sure pool reserves are greater than swap amount
+            if (reserveOut > _amountOut) {
+                sushiTokenIn = UniSushiV2Library.getAmountIn(_amountOut, reserveIn, reserveOut);
+            }
+        }
+
+        // Fails if both the values are maxIn
+        require(!(uniTokenIn == maxIn && sushiTokenIn == maxIn), "ExchangeIssuance: ILLIQUID_SET_COMPONENT");
+        return (uniTokenIn <= sushiTokenIn) ? (uniTokenIn, Exchange.Uniswap, uniswapPair) : (sushiTokenIn, Exchange.Sushiswap, sushiswapPair);
+    }
+
+    /**
+     * Compares the amount of token received for an exact amount of another token across both exchanges,
+     * and returns the max amount.
+     *
+     * @param _amountIn     The amount of input token
+     * @param _tokenA       The address of tokenA
+     * @param _tokenB       The address of tokenB
+     *
+     * @return              The max amount of tokens that can be received across both exchanges
+     * @return              The Exchange on which maximum amount of token can be received
+     * @return              The pair address of the uniswap/sushiswap pool containing _tokenA and _tokenB
+     */
+    function _getMaxTokenForExactToken(uint256 _amountIn, address _tokenA, address _tokenB) internal view returns (uint256, Exchange, address) {
+        if (_tokenA == _tokenB) {
+            return (_amountIn, Exchange.None, ETH_ADDRESS);
```

Comment:
Can someone clarify how the ETH address is being used? Is this just an arbitrary address when the currency token is in the Set?

Reply:
I'm going to ask @ncitron about this quickly. It looks like there's a single place the address could be read:
https://github.com/SetProtocol/index-coop-smart-contracts/blob/7ef60ef3f007de8dd40a9cfcd0c0dd1e0a3b693c/contracts/exchangeIssuance/ExchangeIssuanceV2.sol#L640

Reply:
On slack Noah says: 
> It was originally added so that when we emitted events for the ETH based functions, we could use it as the input/output token value to denote that we issued/redeemed with ETH rather than WETH

> [uncertain] why that got added in the early return for _getMaxTokenForExactToken function. I see no instances when tokenA would even equal tokenB

> We didn't have that check in the original version:

The logic was introduced in: https://github.com/SetProtocol/index-coop-contracts-legacy/pull/13/files

We may have to revisit this and re-deploy with a patch but think this looks ok for staging testing...

Reply:
Sachin confirmed it's ok to ignore this and suggested ETH_ADDRESS could be replaced with ZERO address. 

From slack:

> Regarding the ETH_ADDRESS , when the tokens A and B are same in _getMinTokenForExactToken, we do not need to swap. Which might happen when the input token, say ETH, which is used to acquire other SetToken components, IS one of the SetToken components. So, we need not acquire it from an exchange, and can just use the input tokens provided.

> In that case the return values from `_getMinTokenForExactToken` contains, `Exchange.None`, signifying no exchange, and the pair address should have been ZERO address, but it was a mistake by me, to return ETH_ADDRESS as the pair address there.